### ISSUE TITLE
NPE in StandardPomDependencyManagementConfigurer when a bom that manages a dependency without specifying a version has been imported

### ISSUE THREAD

**Issue by wilkinsona** [spring-gradle-plugins/dependency-management-plugin#213]
Originally reported [here](https://github.com/spring-projects/spring-boot/issues/13193) by @asarkar.

> Gradle 4.7, Boot plugin v2.0.2.RELEASE

> I'm upgrading a bunch of apps from Boot 1.5 to 2, as well as to Gradle 4, and in the process came across the following exception:

>
```
Caused by: java.lang.NullPointerException
    at io.spring.gradle.dependencymanagement.internal.StandardPomDependencyManagementConfigurer.differentVersions(StandardPomDependencyManagementConfigurer.java:175)
    at io.spring.gradle.dependencymanagement.internal.StandardPomDependencyManagementConfigurer.configureBomImports(StandardPomDependencyManagementConfigurer.java:136)
    at io.spring.gradle.dependencymanagement.internal.StandardPomDependencyManagementConfigurer.doConfigurePom(StandardPomDependencyManagementConfigurer.java:115)
```

> There was no indication given that what it was trying to resolve when the NPE occurred. By removing one dependency at a time, I found that it was cause by a Sleuth `mavenBom` import (don't remember which one). The project being an internal one, and because I don't know what caused the NPE, I'm not able to reproduce the problem in a standalone application.

> I think some additional debugging info would be very useful, like what it was trying to resolve when it blew up, and what exactly is null.

**Comment by wilkinsona** [comment]
> I found that it was cause by a Sleuth mavenBom import (don't remember which one)

If you can’t remember which imported bom caused the problem, can you please provide details of all the imported boms? Details of any version overrides would also be useful.

**Comment by asarkar** [comment]
@wilkinsona Thanks for moving the ticket to the right project. Trouble is, I migrated close to 50 projects, and one of them had this issue. If I can find the project, I can find the  BOM, but that's a needle in the haystack. Isn't there anything in the stacktrace to go on?

**Comment by wilkinsona** [comment]
> Isn't there anything in the stacktrace to go on

I can identify what's null from the stacktrace, but, crucially, I can't tell why it's null. The problem avoiding by tolerating null, but the value isn't expected to be null so doing so might just end up masking the root cause of the problem that is elsewhere.

> or perhaps you can add some useful logging as a general improvement?

That logging would only be useful if the problem can be reproduced and it allows some more diagnostics to be captured.

I realise that it won't be easy for you to identify the bom that caused the problem, but without it, it's going to be hard to fix this issue properly. It may be a needle in a haystack for you, but at least you know where the haystack is to start looking.

**Comment by asarkar** [comment]
I _think_ I found the project; following is the scrubbed `build.gradle` before the migration to Boot 2:

```
buildscript {
    repositories {
        mavenCentral()
        maven {
            url 'https://plugins.gradle.org/m2/'
        }
    }
    dependencies {
        classpath(
                "org.springframework.boot:spring-boot-gradle-plugin:$springBootVersion"
        )
    }
}

// more plugins
apply plugin: 'java'
apply plugin: 'org.springframework.boot'

allprojects {
    group = 'com.mycompany.cloud'
    project.version = scmVersion.version

    apply from: "$gradleCommonSettings/java-settings/build.gradle"
    apply from: "$gradleCommonSettings/spring-cloud-settings/build.gradle"

    // more stuff

    dependencies {
        compile(
                'org.springframework.cloud:spring-cloud-starter-feign',
                'org.springframework.cloud:spring-cloud-starter-ribbon',
                'org.springframework.cloud:spring-cloud-starter',
                'org.springframework.cloud:spring-cloud-starter-hystrix',
        )

        runtime(
                "org.springframework.cloud:spring-cloud-starter-sleuth",
                "org.springframework.cloud:spring-cloud-sleuth-zipkin"
        )

        testCompile("org.springframework.boot:spring-boot-starter-test:$springBootVersion")
    }

    apply from: "$gradleCommonSettings/maven-publish-settings/build.gradle"
    task deploy(dependsOn: [install, uploadArchives])
}

dependencies {
    compile(
            "com.mycompany.cloud:auth-client:$departmentApiVersion",
            'org.springframework.cloud:spring-cloud-starter-zuul',
            "com.mycompany.content:department-api:$departmentApiVersion",
            "com.mycompany.content:department-context-support:$departmentApiVersion",
            "com.mycompany.content:department-content-support:$departmentApiVersion",
            "com.mycompany.content:department-logging:$departmentApiVersion",
            "io.vavr:vavr:$vavrVersion",
            "org.springframework.boot:spring-boot-starter-actuator",
            'org.springframework.cloud:spring-cloud-starter-config',
            'org.springframework.boot:spring-boot-starter-web',
    )
    runtime("org.springframework.cloud:spring-cloud-starter-sleuth:$springSleuthVersion",
            "org.springframework.cloud:spring-cloud-sleuth-zipkin:$springSleuthVersion",
            "com.mycompany.content:department-metric-support:$departmentApiVersion"
    )
    // more
}
```
The variable values come from `gradle.properties`, and the ones relevant to are as follows:

springBootVersion=1.5.3.RELEASE
springSleuthVersion=1.2.0.RELEASE

**Comment by wilkinsona** [comment]
Thanks for taking the time to find that. I can’t see any boms being imported so I assume that’s happening in some of your common settings. Can you share the relevant parts of those too please? I’d like to be able to piece together something that reproduces the problem.

**Comment by asarkar** [comment]
@meidea Not sure what are you saying.

**Comment by meidea** [comment]
@asarkar include mavenBom 'org.springframework.cloud:spring-cloud-sleuth:2.0.0.RC1'  will cause the   problem

**Comment by asarkar** [comment]
@meidea thanks; If you've a sample app, that'll surely help.

**Comment by wilkinsona** [comment]
Thanks, @meidea. That looks to me to be due to a mistake in `spring-cloud-sleuth` has it's managing a dependency without actually specifying a version. That said, this plugin should tolerate such an unusual bom so this issue can be used to do that.

**Comment by semaphore3000** [comment]
The problem also exists from versions 1.5.10.RELEASE to 1.5.15.RELEASE.
So could you please fix it in the 1.5.x branch?

**Comment by wilkinsona** [comment]
Spring Boot 1.5.15 has already upgraded to Dependency Management Plugin 1.0.6 which should contain this fix. If you believe that's not the case, please open a new issue with a minimal sample that reproduces the problem.
